lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -1185,11 +1185,13 @@ class DedupPass : public DedupBase<DedupPass> {
     DenseMap<Attribute, StringAttr> dedupMap;
 
     // We must iterate the modules from the bottom up so that we can properly
-    // deduplicate the modules. We use early increment so we can safely delete
-    // nodes from the instance graph as we iterate through it.
-    for (auto *node :
-         llvm::make_early_inc_range(llvm::post_order(&instanceGraph))) {
-      auto module = cast<FModuleLike>(*node->getModule());
+    // deduplicate the modules. We copy the list of modules into a vector first
+    // to avoid iterator invalidation while we mutate the instance graph.
+    std::vector<FModuleLike> modules;
```

Comment:
Any reason for `std::vector<FModuleLike>` over the default-sized `SmallVector<FModuleLike>`?

Reply:
FWIW my reading of the guide is `SmallVector<FModuleLike>` is generally preferred over `std::vector<FModuleLike>`, even if the vector isn't necessarily small. I'm pretty sure `SmallVector<FModuleLike>` just means `SmallVector<FModuleLike, 0>`.

Reply:
@seldridge missed your last comment, I will look into it.

> FWIW my reading of the guide is SmallVector<FModuleLike> is generally preferred over std::vector<FModuleLike>

Yeah, the guide is saying in pretty uncertain terms that `std::vector` has fallen out of favor. I'll switch it back when I see if I can do what Schuyler suggested.

> I'm pretty sure SmallVector<FModuleLike> just means SmallVector<FModuleLike, 0>

I think they're actually different, `N=0` forces the storage out of line, the [default value](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/SmallVector.h#L1173-L1185) of `N` is calculated based on the size of what you're storing but its [at least one inline element](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/SmallVector.h#L1123-L1127)

Reply:
I was curious how often each is actually used, just grepping for occurances:
|  | MLIR | LLVM |
|---|---|---|
| `std::vector`| 951 | 22148 |
| `SmallVector<x, 0>`| 43 | 439 |

Reply:
Unfortunately I still needed the cast, so I used a `map_range` to coerce it to the correct thing.